text/3502-cargo-script.md

```diff
@@ -0,0 +1,1229 @@
+
+While `rust-analyzer` needs to be able to distinguish regular `.rs` files from
+single-file packages to look up the relevant manifest to perform operations, we
+propose that be through checking the `#!` line (e.g.
+[how perl detects perl in the `#!`](https://stackoverflow.com/questions/38059830/how-does-perl-avoid-shebang-loops).
+While this adds boilerplate for Windows developers, this helps encourage
+cross-platform development.
+
+If we adopted a unique file extensions, some options include:
+- `.crs` (used by `cargo-script`)
+- `.ers` (used by `rust-script`)
+  - No connection back to cargo
+- `.rss`
+  - No connection back to cargo
+  - Confused with RSS
+- `.rsscript`
+  - No connection back to cargo
+  - Unwieldy
+- `.rspkg`
+  - No connection back to cargo but conveys its a single-file package
+
+## Embedded Manifest Format
+
+Considerations for embedded manifest include
+- How obvious it is for new users when they see it
+- How easy it is for newer users to remember it and type it out
+- How machine editable it is for `cargo add` and friends
+- Needs to be valid Rust code based on the earlier stated design guidelines
+- Lockfiles might also need to reuse how we attach metadata to the file
+
+See [RFC 3503](https://github.com/rust-lang/rfcs/pull/3503) for discussion on syntax.
+
+The `cargo` infostring was chosen because
+- An alternative is `Cargo.toml` but infostring language fields are generally an identifier rather than a file name
+- An alternative is to specify the cargo payload as an attribute, like `cargo,file=Cargo.toml` but the language should generally convey the format/schema rather than an attribute
+  - This also adds a lot of verbosity to teach, write, and get wrong
+- If we add additional frontmatter blocks in the future (like for lockfiles),
+  the manifest block will be needed at minimum and other blocks will likely
+  only be used in 1% of cases the manifest block is present, so we shouldn't
+  hurt the simple, common case for the more advanced, long tail cases.
+
+## `edition`
+
+[The `edition` field controls what variant of cargo and the Rust language to use to interpret everything.](https://doc.rust-lang.org/edition-guide/introduction.html)
+
+A policy on this needs to balance
+- Matching the expectation of a reproducible Rust experience
+- Users wanting the latest experience, in general
+- Boilerplate runs counter to experimentation and prototyping, particularly in the "no dependencies" case
+  - A `cargo new --script` (flag TBD) could help reduce writing of boilerplate.
+- There might not be a backing file if we read from `stdin` (future possibility)
+
+**Solution: Latest as Default**
+
+Default to the `edition` for the current `cargo` version, assuming single-file
+packages will be transient in nature and users will want the current `edition`.
+However, we will produce a warning when no `edition` is specified, nudging
+people towards reproducible code.
```

Comment:
I'm wondering about the term "warning" here. This seems more like an informational thing to me. If we use "warning" then the user may feel like they did something wrong and needs to fix it, even if it runs just fine for the prototyping work they are doing.

Reply:
unrelated to the above, is the compiler clever enough to tell if a compilation error happened because the wrong edition was used? if that was the case we could potentially only show the warning if something actually fails

Reply:
We are walking a fine line here and I think "warning" level is appropriate.  Once we have rust-lang/cargo#12235, "diagnostic" might be a more appropriate word which could be fixed with `cargo fix`.


We are trying to balance
- Low boilerplate (i.e. don't make the frontmatter required)
- Behave as the user would expect (i.e. the regular 2015 default would likely be confusing and would just force people into using a frontmatter)
- Maintain the aims of the edition system, particularly that code doesn't break

In a lot of the previous discussions, some have voiced concerns that a warning might not be enough and I'm surprised it hasn't come up here yet (for any who feel this way, I'd recommend trying it out!).

Warnings are fine for prototyping (there is a reason people generally recommend against `#![deny(warnings)]` but instead recommend `RUSTFLAGS=-Dwarnings`) but the fact that we might break people's code in upgrade is something they really need to be aware of and cautious about, hence a warning.